Fixed solveset and added tests for already solved issues

oscargus opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Closes #10085 (added test, solved elsewhere)
Closes #10876 (added test, solved elsewhere)
Closes #12032 (added test, solved elsewhere)
Fixes #13396 (solved underlying problem)
Added test for #13961 (gives two identical solutions)
Closes #14541 (added test, solved elsewhere)
Closes #14642 (added test, solved elsewhere)
Added test for #16643 (original issue solved, but not follow-up equation)
Fixes #17580 (solved underlying problem)
Closes #17906 (added test, solved elsewhere(?))
Fixes #17940 (solved underlying problem)
Fixes #19506 (solved underlying problem)



#### Brief description of what is fixed or changed
Added tests and closed where appropriate.

#### Other comments

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* solvers
    * `solveset` correctly solves `arg` when the argument is complex.

<!-- END RELEASE NOTES -->

oscargus committed:
Added tests for #10085, #10876, #12032, #13396, #13961, #14541, #14642

oscarbenjamin commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1724,6 +1731,130 @@ def test_solve_nonlinear_trans():
 
 
+def test_issue_14642():
+    x = symbols('x',dtype = complex)
```
What does `dtype=complex` do?

oscargus replied:
No idea. The original code was like this...

oscarbenjamin replied:
I guess it is supposed to be `complex=True` or something. It should be removed though because it doesn't do anything (ideally this would raise an error).

oscargus replied:
OK! Will do!

oscargus commented:
I've updated the tests and original comment so that the status is correct.

oscarbenjamin commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1724,6 +1731,150 @@ def test_solve_nonlinear_trans():
+def test_issue_13396_failing():
+    expr = -2*y*exp(-x**2 - y**2)*Abs(x)
+    # The second part may be empty
+    sol = Union(FiniteSet(Integer(0)), Intersection(S.Reals, FiniteSet(Pow(Add(Mul(Integer(-1),
+                Pow(x, Integer(2))), zoo), Rational(1, 2)), Mul(Integer(-1),
+                 Pow(Add(Mul(Integer(-1), Pow(x, Integer(2))), zoo), Rational(1, 2))))))
+
+     # Should work, but will not. See #21943
+    assert solveset(expr, y, domain=S.Reals) == sol
```
Problem is that the args are flipped in `zoo - x**2`.

oscarbenjamin replied:
I guess that the *cause* of the problem is evaluate=False being used somewhere...

oscargus replied:
I tried to rewrite the expressions some time ago, but couldn't figure out where to put the `evaluate=False`. May have another go, but indeed it would be better to only get one solution...

oscarbenjamin replied:
To recreate the output of `solveset` we need `evaluate=False` for the `Add` in `zoo - x**2`:
```python
In [15]: sol = Union(FiniteSet(Integer(0)), Intersection(S.Reals, FiniteSet(Pow(Add(Mul(Integer(-1),
    ...:             Pow(x, Integer(2))), zoo, evaluate=False), Rational(1, 2)), Mul(Integer(-1),
    ...:             Pow(Add(Mul(Integer(-1), Pow(x, Integer(2))), zoo, evaluate=False), Rational(1, 2))))))
    ...: 
    ...: expr = -2*y*exp(-x**2 - y**2)*Abs(x)
    ...: 
    ...: print(solveset(expr, y, Reals) == sol)
True
```

oscarbenjamin replied:
It's the 2-arg Mul:
```python
In [6]: e = zoo - x**2

In [7]: e.args
Out[7]: 
⎛       2⎞
⎝zoo, -x ⎠

In [8]: e2 = -1 * e

In [9]: e2
Out[9]: 
 2      
x  + zoo

In [10]: e2.args
Out[10]: 
⎛ 2     ⎞
⎝x , zoo⎠

In [11]: e2 == e2.func(*e2.args)
Out[11]: False
```
It can be fixed with
```diff
diff --git a/sympy/core/mul.py b/sympy/core/mul.py
index 9933eb601a..960b5baf25 100644
--- a/sympy/core/mul.py
+++ b/sympy/core/mul.py
@@ -281,7 +281,7 @@ def flatten(cls, seq):
                         ar = a*r
                         if ar:
                             bargs.insert(0, ar)
-                        bargs = [Add._from_args(bargs)]
+                        bargs = [Add(*bargs)]
                         rv = bargs, [], None
             if rv:
                 return rv
```

smichr replied:
or maybe `_unevaluated_Add(*args)` -- I read the issue but am not seeing enough context to understand the problem. `Add._from_args` is much faster than `Add` and since you are making this change in core it makes me nervous.

smichr replied:
> It can be fixed with

I think this is best. Otherwise it's dicey with special handling of objects in Add which get promoted to the coefficient position (like TensExpr, AccumBounds, ComplexInfinity).

And there is a difference between the handling of `-x` vs `-1*x`. The following diff will make the behavior more uniform:
```diff
diff --git a/sympy/core/mul.py b/sympy/core/mul.py
index 9933eb601a..1de76341f1 100644
--- a/sympy/core/mul.py
+++ b/sympy/core/mul.py
@@ -267,7 +267,7 @@ def flatten(cls, seq):
                 r, b = b.as_coeff_Mul()
                 if b.is_Add:
                     if r is not S.One:  # 2-arg hack
-                        # leave the Mul as a Mul?
+                        # 2*Mul(3, x + 1, evaluate=False) -> 6*(x + 1)
                         ar = a*r
                         if ar is S.One:
                             arb = b
@@ -275,14 +275,12 @@ def flatten(cls, seq):
                             arb = cls(a*r, b, evaluate=False)
                         rv = [arb], [], None
                     elif global_parameters.distribute and b.is_commutative:
-                        r, b = b.as_coeff_Add()
-                        bargs = [_keep_coeff(a, bi) for bi in Add.make_args(b)]
-                        _addsort(bargs)
+                        r, nonr = b.as_coeff_Add()
+                        bargs = [_keep_coeff(a, bi) for bi in Add.make_args(nonr)]
                         ar = a*r
                         if ar:
                             bargs.insert(0, ar)
-                        bargs = [Add._from_args(bargs)]
-                        rv = bargs, [], None
+                        rv = [Add(*bargs)], [], None
             if rv:
                 return rv
 
@@ -2144,4 +2142,4 @@ def do(e):
 
 from .numbers import Rational
 from .power import Pow
-from .add import Add, _addsort, _unevaluated_Add
+from .add import Add, _unevaluated_Add
diff --git a/sympy/core/tests/test_arit.py b/sympy/core/tests/test_arit.py
index 16580c14fa..8a35d99a49 100644
--- a/sympy/core/tests/test_arit.py
+++ b/sympy/core/tests/test_arit.py
@@ -2354,3 +2354,24 @@ def test_issue_17130():
 def test_issue_21034():
     e = -I*log((re(asin(5)) + I*im(asin(5)))/sqrt(re(asin(5))**2 + im(asin(5))**2))/pi
     assert e.round(2)
+
+
+def test_issue_22021():
+    from sympy.calculus.util import AccumBounds
+    from sympy.tensor.tensor import TensExpr
+    from sympy.utilities.iterables import permutations
+    # these objects are special cases in Mul
+    from sympy.tensor.tensor import TensorIndexType, tensor_indices, TensorSymmetry,    get_symmetric_group_sgs, TensorIndex, tensor_mul, TensAdd,tensor_heads
+    L = TensorIndexType("L")
+    i = tensor_indices("i", L)
+    A, B = tensor_heads("A B", [L])
+    e = A(i) + B(i)
+    assert -e == -1*e
+    e = zoo + x
+    assert -e == -1*e
+    a = AccumBounds(1, 2)
+    e = a + x
+    assert -e == -1*e
+    for args in permutations((zoo, a, x)):
+        e = Add(*args, evaluate=False)
+        assert -e == -1*e
```

oscargus commented:
I added the `evalute=False` in the test with a comment about that they should ideally not be required.

Didn't want to change the core mul-code...

oscargus committed:
Fixed solveset issues, added tests and bit of clean-up

oscargus commented:
I have fixed the `zoo` problems in a slightly more general way than discussed in #22019.

Also fixed solving `arg` in a way that at least is correct, but may not always return "minimal" solutions. For example, `solveset(arg(x) , x)` returns

`Intersection(ConditionSet(x, Eq(im(x), 0), Complexes), ConditionSet(x, re(x) > 0, Complexes))`

where the first ConditionSet is identical to x being real.

On the other hand, `solveset(arg(x-I) , x)` now do not throw and error, but returns

`Intersection(ConditionSet(x, Eq(im(x) - 1, 0), Complexes), ConditionSet(x, re(x) > 0, Complexes))`

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -197,6 +197,11 @@ def invert_real(f_x, y, x, domain=S.Reals):
+    # Some inverse functions evaluate to zoo, remove if present
+    fs = FiniteSet(S.ComplexInfinity)
+    if fs.is_subset(g_ys):
+        g_ys -= fs
```
I'm not sure this does fix things in a more general way. It removes zoo but loses the reason for removing zoo.

oscargus replied:
Well, the option is to check each and every inverse function that the input will not lead to zoo as you suggested. I cannot really see that keeps the reason better.

oscargus replied:
My point is that once the function class is identified it will in almost all cases call `_invert_real` again with an evaluated version of the inverse function. So rather than checking for every evaluation if the input may yield `zoo`, which is a bit of a hassle and cases will have to be added once identified, it seems more general to check if the input to (the next call of) `_invert_real` is `zoo`. The only possible way that this may not work is if there are some functions that map `zoo` back to reals in a "controlled" way (not just inverting again). It doesn't seem to be any of those cases in the tests so far.

I added a test when I wrote this code to illustrate the point:
`assert solveset(atan(x**2 - y**2)-pi/2, y, S.Reals) == EmptySet()` which currently returns `Intersection({-sqrt(x**2 + zoo), sqrt(x**2 + zoo)}, Reals)`.

oscarbenjamin replied:
I think it would be better if `invert_real` only returned sets that are strictly contained in the domain of a function. There are other ways this can go wrong like if if the `subset` test is indeterminate then a zoo case is not removed.

smichr replied:
> I think it would be better if `invert_real` only returned sets that are strictly contained in the domain of a function

Since it has the `domain` argument, I agree. I think I fixed a problem where the domain was integers but it wasn't return only integers. Those are troublesome bugs because you assume that the routine can be counted on to respect the domain and that you won't have worry about this in the the calling routine.

oscargus replied:
I agree in general. The only other way I might think of is 
```
g_ys = Intersection(g_ys, S.Reals)
```
which will also cover other cases (however, I believe that all the inverse functions return real-valued results for real-valued input, except for those that returns zoo, but I may be wrong). I'm a bit worried that this could explode in certain cases. (`Intersection(ImageSet(..., Intersection(ImageSet(..., Intersection(...)), S.Reals), S.Reals)`) Also, I am not sure how computationally efficient this is, but can give it a try.

oscargus replied:
Will also be interesting see if that deals with: `solveset(I*pi-log(x), x, S.Reals)`, which, if I guess correctly, will have an imaginary intermediate value.

smichr replied:
I tried to fix this at the intersection-level in #22045.

smichr commented:
Hmmm. The signature of `invert_real` should be like `solveset_real` and not accept the domain argument per its own docstring

```python
def invert_real(f_x, y, x):
    """
    Inverts a real-valued function. Same as _invert, but sets
    the domain to ``S.Reals`` before inverting.
    """
    return _invert(f_x, y, x, S.Reals)
```
And then the calling program intersects the real result with the actual domain.

Or else it should be something like this:

```python
def invert_real(f_x, y, x, domain=S.Reals):
    """
    Inverts a real-valued function. Same as _invert, but sets
    the domain to ``S.Reals`` before inverting.
    """
    return _invert(f_x, y, x, domain.intersection(S.Reals))
```

smichr commented:
```diff
diff --git a/sympy/solvers/solveset.py b/sympy/solvers/solveset.py
index acb0a81052..434520fd79 100644
--- a/sympy/solvers/solveset.py
+++ b/sympy/solvers/solveset.py
@@ -188,10 +188,10 @@ def _invert(f_x, y, x, domain=S.Complexes):
 
 def invert_real(f_x, y, x, domain=S.Reals):
     """
-    Inverts a real-valued function. Same as _invert, but sets
-    the domain to ``S.Reals`` before inverting.
+    Inverts a real-valued function. Same as _invert, but
+    limits the domain to ``S.Reals`` before inverting.
     """
-    return _invert(f_x, y, x, domain)
+    return _invert(f_x, y, x, domain.intersection(S.Reals))
 
 
 def _invert_real(f, g_ys, symbol):
diff --git a/sympy/solvers/tests/test_solveset.py b/sympy/solvers/tests/test_solveset.py
index fbbc6cddae..3ff2094bf0 100644
--- a/sympy/solvers/tests/test_solveset.py
+++ b/sympy/solvers/tests/test_solveset.py
@@ -65,6 +65,7 @@ def ireal(x, s=S.Reals):
     assert invert_real(x, 0, x, Interval(1, 2)) == (x, S.EmptySet)
 
     assert invert_real(exp(x), z, x) == (x, ireal(FiniteSet(log(z))))
+    assert invert_real(exp(x), z, x, S.Complexes) == (x, ireal(FiniteSet(log(z))))
 
     y = Symbol('y', positive=True)
     n = Symbol('n', real=True)
```
passes all solveset tests

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -197,6 +197,11 @@ def invert_real(f_x, y, x, domain=S.Reals):
 def _invert_real(f, g_ys, symbol):
     """Helper function for _invert."""
 
+    # Some inverse functions evaluate to zoo, remove if present
+    fs = FiniteSet(S.ComplexInfinity)
+    if fs.is_subset(g_ys):
+        g_ys -= fs
+
```
```suggestion
```

oscargus committed:
Remove introduced code.

Co-authored-by: Christopher Smith <smichr@gmail.com>

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -1116,7 +1118,16 @@ def _solveset(f, symbol, domain, _check=False):
+            if domain is S.Complexes:
+                return result
+            # Add domain
+            new_res = Intersection(result, domain)
+            # Check if the domain is really contributing
+            if new_res.is_subset(result):
```
Isn't this always True? (Or rather is it not safe to assume that it is always True?)

oscargus replied:
It may be that you are correct.

The thing was that just adding the domain sometimes led to some bad looking results. Typically, the domain was not really required in some cases and still remained. I do not remember an exact example, but, say, `Intersection(ConditionSet(... S.Reals), S.Reals)` or something similar. This was one of many attempts to get rid of those that at least didn't add anything "stupid".

Maybe it would be better to just add the domain unless it is `S.Complexes` and use that as input for improving `Intersection`?

oscargus replied:
Ahh, forgot what I had already written. I'll try to find a better way of doing this.

oscargus replied:
This was the problem:
`solveset(Abs(x) - n, x, S.Reals)` currently returns `ConditionSet(x, Contains(n, Interval(0, oo)), {-n, n})`, by Intersecting with the Reals domain we get `ConditionSet(x, Contains(_n, Interval(0, oo)), Intersection({-_n, _n}, Reals))`.

oscargus replied:
I realize that I was trying to solve issues that I didn't fully understood. So I'll just revert this part.
